Replace proxy layer with ProxyManager's layer

Ocramius opened the pull request:
Replaces #1241

TODOs:

 * [x] replace proxy instantiation with ProxyManager proxies
 * [x] ensure that `ProxyManager` class names are properly recognized
 * [x] remove `ClassUtils` shim
 * [x] ~~~move proxy instantiation and definition to the metadata~~~
 * [x] remove `Proxy*` related code in ORM
 * [x] update `Doctrine\ORM\Configuration` to work with `ProxyManager` utilities
 * [x] update `ClassUtils` to work with `ProxyManager` proxy class names
 * [x] remove public properties hack
 * [x] extract identifier normalization to its own object that is the opposite of [`IdentifierFlattener`](https://github.com/doctrine/doctrine2/blob/1d5c87ee4bcf22d699530b3792739f33b91cf6cd/lib/Doctrine/ORM/Utility/IdentifierFlattener.php)
 * [x] test "friend object" behavior with proxies (shared private scope access)
 * [x] test that `EntityManager#getReference()` calls `ProxyFactory#getProxy()` with un-touched data
 * [x] document BC break extent:
    * [x] `__sleep` behavior changes
    * [x] `__wakeup` behavior changes
    * [x] `serialize()` no longer breaks the graph -> potentially dangerous if serializing over large association graphs
    * [x] `Proxy` interface gone
    * [x] `private` property access now triggers lazy-loading
    * [x] `ReflectionProperty` property access now triggers lazy-loading
 * [x] document improvements
    * [x] "friend object" behavior with proxies (shared private scope access)
    * [x] `private` property access now triggers lazy-loading
    * [x] `ReflectionProperty` property access now triggers lazy-loading
    * [x] proxies no longer lazy-load when no internal state is accessed
    * [x] no more "identifier getter" hacks
    * [x] no more type conversion hacks
    * [x] identifier fields are always set, and always have the expected type
 * [x] benchmark regression testing

Ocramius committed:
Stubbing out a retry on the s/`Proxy`/`GhostObjectInterface` replacement patch

Ocramius committed:
Ensuring that the `BasicFunctionalTest` relies on the `GhostObjectInterface` API

Ocramius committed:
Prevent setting any values in un-initialized proxies

Ocramius committed:
Proxy identifier must be set when the proxy is instantiated

Ocramius committed:
Correcting `ProxyFactoryTests`, which were opinionated around past proxy definition

Ocramius committed:
Fixing faulty test that was expecting a different entity to bbe hydrated during proxy initialization

Ocramius committed:
Correcting reference to non-existing method

Ocramius committed:
ORM proxy names may be looked up in the class metadata factory - the parser should not manually do `class_exists` checks unless strictly required

Ocramius committed:
Hardening `ProxiesLikeEntitiesTest` by adding some precondition checks

Ocramius committed:
Cascade operations do not apply to non-initialized proxies: no associations are set there

Ocramius committed:
Transient fields should be skipped by proxy initialization semantics

Ocramius committed:
Fixed reference to non-existing method

Ocramius committed:
Cleaning up `ReferenceProxyTest`, which no longer reflects current invariants

Ocramius committed:
Correcting `DDC1193Test`, which was using the old `Proxy` API

Ocramius committed:
Corrected `DDC1228Test`, which was quite broken and relying on the old `Proxy` API

Ocramius committed:
Correcting `DDC1690Test`: even if a proxy is not yet loaded, it must have notify property listeners to function correctly

Ocramius committed:
The `DefaultQueryCache` still contained `instanceof Proxy` checks - replaced with `GhostObjectInterface` checks

Ocramius committed:
Correcting `DDC2231Test`, since proxies no longer initialize if no state is mutated

Ocramius committed:
Hardening DDC3223Test with additional pre- and post-conditions

Ocramius committed:
Removing DDC-1238 related code. Issue seems to be invalid, as cloning a proxy should NOT modify its state, but just make it non-managed

Ocramius committed:
Documenting DDC-1022 tests to be removed, as proxy behavior changed around `__clone` and `__wakeup` semantics

Ocramius committed:
Correcting DDC-1238 to remove non-sensical semantics that cleared proxy data after detach operations

Ocramius committed:
Correcting metadata factory tests - fetching a class by name does indeed fetch the class even if prefixed

Ocramius committed:
Adding `ocramius/proxy-manager` dependency

Ocramius committed:
`SecondLevelCacheManyToOneTest` should not cause more queries for fetching associated identifiers (which are already available)

Ocramius committed:
Adding stub implementation of logic that converts a flat identifier into a deep identifier

Ocramius committed:
Explicit BC break: we no longer call `__wakeup` on proxies

Ocramius committed:
Correcting `ReferenceProxyTest`: proxy identifiers are no longer unset while cloning, but they are not considered managed

Ocramius committed:
Stop registering cloned proxies (real instance being managed) as managed

Ocramius committed:
The `NotifyPropertyChanged` API should only be used once we decided to overwrite the entity data, in which case the `UnitOfWork` is indeed tracking entity state

Ocramius commented on lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php:
```diff
@@ -124,11 +124,11 @@ protected function hydrateRowData(array $sqlResult, array &$result)
-
-            $this->em->getUnitOfWork()->registerManaged($this->hints[Query::HINT_REFRESH_ENTITY], $id, $data);
-        }
+//        if (isset($this->hints[Query::HINT_REFRESH_ENTITY])) {
```
This code was extremely broken, but I still need to see if any other related code may have been broken by it.

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/DefaultProxyResolver.php:
```diff
@@ -5,9 +5,6 @@
-use Doctrine\Common\Util\ClassUtils;
-use Doctrine\ORM\Proxy\Proxy;
-
 class DefaultProxyResolver implements ProxyResolver
```
To be removed

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/ProxyDefinitionFactory.php:
```diff
@@ -58,11 +58,12 @@ public function build(ClassMetadata $classMetadata) : ProxyDefinition
-
-            $this->generatorStrategy->generate($proxyClassPath, $definition);
-        }
+        // @TODO kill this. We no longer generate proxies the "old way"
```
To be removed

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/ProxyFactory.php:
```diff
@@ -56,11 +57,6 @@ public function generateProxyClasses(array $classMetadataList) : int;
-     * @return \Doctrine\ORM\Proxy\Proxy
      */
-    public function getProxy(string $className, array $identifier) : Proxy;
+    public function getProxy(string $className, array $identifier) : GhostObjectInterface;
```
Should probably become `ClassMetadata $class` to avoid repeating metadata lookup operations

Ocramius replied:
Not changing this for now - can be done later, eventually

guilhermeblanco replied:
Do it now and make it fully finished! Unless you plan to make subsequent PRs on the funcionality.

Ocramius replied:
I'll dig in today 👍

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/ProxyGenerator.php:
```diff
@@ -6,7 +6,7 @@
 namespace Doctrine\ORM\Proxy\Factory;
```
To be removed

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -7,8 +7,13 @@
 
 use Doctrine\ORM\Configuration\ProxyConfiguration;
```
(probably) to be removed

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -82,18 +87,110 @@ public function generateProxyClasses(array $classMetadataList) : int
+        $proxyInstance = (new LazyLoadingGhostFactory(new Configuration()))
+            ->createProxy(
+                $metadata->getClassName(),
+                function (
```
This loader should be cached

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -82,18 +87,110 @@ public function generateProxyClasses(array $classMetadataList) : int
     {
+        $metadata = $this->entityManager->getClassMetadata($className);
+
+        $proxyInstance = (new LazyLoadingGhostFactory(new Configuration()))
```
This factory should be injected

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -82,18 +87,110 @@ public function generateProxyClasses(array $classMetadataList) : int
+                },
+                [
+                    // @TODO this should be a constant reference, not a magic constant
+                    'skippedProperties' => \array_merge(
```
For now, we'll have to use a `private const`

Ocramius commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -965,11 +969,25 @@ public function AbstractSchemaName()
     {
-        if (! (class_exists($schemaName, true) || interface_exists($schemaName, true))) {
-            $this->semanticalError("Class '$schemaName' is not defined.", $this->lexer->token);
+        // @TODO are these checks even needed?
```
The check should probably be inversed: `getClassMetadata()` first, then `(class|interface)_exists()`. The reason being that `class_exists()` can trigger I/O, while `getClassMetadata()` usually doesn't.

This makes `SELECT new Foo(b.c) FROM Bar b` slower, as `getClassMetadata()` fails more often, but it makes other DQL queries faster.

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2108,29 +2112,31 @@ public function createEntity($className, array $data, &$hints = [])
-                    $property = $class->getProperty($fieldName);
+                // @TODO DDC-1238 seems to be invalid.
+                // @TODO We shouldn't touch the object state, but instead just consider it "unmanaged" and skip further hydration
+//                foreach ($class->identifier as $fieldName) {
```
This block was super-broken: we just want an un-managed instance, but we really don't want to change its identifier. Replacing the identifier = bad.

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2108,29 +2112,31 @@ public function createEntity($className, array $data, &$hints = [])
+                && $unmanagedProxy instanceof GhostObjectInterface
                 && $this->isIdentifierEquals($unmanagedProxy, $entity)
             ) {
+                // @TODO messy, but for now we try continuing operation as normal
```
No longer valid once removing this block

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2108,29 +2112,31 @@ public function createEntity($className, array $data, &$hints = [])
                 && $this->isIdentifierEquals($unmanagedProxy, $entity)
             ) {
+                // @TODO messy, but for now we try continuing operation as normal
+                // @TODO ideally, all this scenario should be completely gone
```
No longer valid once removing this block

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2108,29 +2112,31 @@ public function createEntity($className, array $data, &$hints = [])
                 // refreshed object may be anything
-                foreach ($class->identifier as $fieldName) {
-                    $property = $class->getProperty($fieldName);
+                // @TODO DDC-1238 seems to be invalid.
```
This refers to the [`DDC1238Test`](https://github.com/doctrine/doctrine2/blob/1d5c87ee4bcf22d699530b3792739f33b91cf6cd/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC1238Test.php)

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2108,29 +2112,31 @@ public function createEntity($className, array $data, &$hints = [])
-                }
-
-                return $unmanagedProxy;
+                // @TODO overall, this logic seems wrong and harmful. Even if the proxy is un-managed, we want to load its data here
```
No longer valid

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2108,29 +2112,31 @@ public function createEntity($className, array $data, &$hints = [])
-
-                return $unmanagedProxy;
+                // @TODO overall, this logic seems wrong and harmful. Even if the proxy is un-managed, we want to load its data here
+//                return $unmanagedProxy;
```
No longer valid

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2158,17 +2164,17 @@ public function createEntity($className, array $data, &$hints = [])
             return $entity;
         }
 
+        if ($entity instanceof NotifyPropertyChanged) {
```
This was moved, because the `addPropertyChangedListener()` call must happen both for initialized and un-initialized proxies once the `UnitOfWork` starts touching them

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2339,28 +2345,50 @@ public function createEntity($className, array $data, &$hints = [])
                     // Proxies do not carry any kind of original entity data until they're fully loaded/initialized
                     $managedData = [];
 
+                    // @TODO refactor: `getProxy`: no scalars if the identifier is an association, use real value
```
Comment block to be removed: replicated below

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2339,28 +2345,50 @@ public function createEntity($className, array $data, &$hints = [])
                             break;
                     }
 
+                    // @TODO using `$associatedId` here seems to be risky.
```
Very risky indeed: this should be a `flattenId(normalizeId(flatId))` instead, in order to guarantee consistency

Ocramius replied:
This code cannot be fixed in this patch: possibly related to #6638, #6648, #6649

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2369,7 +2397,16 @@ public function createEntity($className, array $data, &$hints = [])
             $association->setValue($entity, $newValue);
 
-            if ($association->getInversedBy() && $association instanceof OneToOneAssociationMetadata) {
+            if (
```
I think this entire block seems wrong and should be removed. The code in this block overwrites the association field of the related entity, which is wrong.

Ocramius replied:
This code cannot be fixed in here

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2531,6 +2568,86 @@ public function getSingleIdentifierValue($entity)
+     *
+     * @return GhostObjectInterface|object
+     */
+    private function tryGetByIdOrLoadProxy(array $id, string $rootClassName)
```
This method is extracted from the `EntityManager` API

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2531,6 +2568,86 @@ public function getSingleIdentifierValue($entity)
+     *
+     * @return mixed[]
+     */
+    private function convertFlatIdentifierIntoRealIdentifierFieldValues(
```
This method is to be extracted to the `ClassMetadata` API, which should (given a `UnitOfWork`) normalize/denormalize identifiers for us.

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2531,6 +2568,86 @@ public function getSingleIdentifierValue($entity)
+                $targetIdMetadata = $this->em->getClassMetadata($declaredProperty->getTargetEntity());
+
+                $normalizedAssociatedId[$name] = $this->tryGetByIdOrLoadProxy(
+                    [reset($targetIdMetadata->identifier) => $flatIdentifier[$name]], // @TODO this is where stuff breaks with composite IDs :-(
```
This assumes single-field identifiers on associated proxies: extremely dangerous

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/ProxiesLikeEntitiesTest.php:
```diff
@@ -52,6 +59,8 @@ protected function setUp()
         $this->em->flush();
         $this->em->clear();
+
+        $this->proxyClassName = (new Configuration())->getClassNameInflector()->getProxyClassName(CmsUser::class);
```
This is probably not the correct approach

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/ReferenceProxyTest.php:
```diff
@@ -165,46 +167,30 @@ public function testInitializeChangeAndFlushProxy()
-    /**
-     * @group DDC-1022
-     */
-    public function testWakeupCalledOnProxy()
```
This no longer holds true: we don't call `__wakeup` on proxies anymore.

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC1238Test.php:
```diff
@@ -58,14 +59,29 @@ public function testIssueProxyClear()
+            'Even if a proxy is detached, it should still have an identifier'
+        );
+
+        $user2->initializeProxy();
```
Just added a block checking that we didn't ruin the fresh instance.

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC3223Test.php:
```diff
@@ -67,7 +81,11 @@ class Participant
-    /** @ORM\ManyToOne(targetEntity="ProfileStatus") */
+    /**
+     * @ORM\ManyToOne(targetEntity="ProfileStatus")
+     * @ORM\JoinColumn(name="status_id", nullable=false)
```
Added the `nullable=false` since I added the `@var` block with no nullability

Ocramius commented on tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php:
```diff
@@ -185,40 +191,42 @@ public function testFailedProxyCloningDoesNotMarkTheProxyAsInitialized()
 
     public function testProxyClonesParentFields()
     {
-        $companyEmployee = new CompanyEmployee();
```
This was completely broken: the persister was returning a new instance instead of just hydrating the given object.

Ocramius committed:
Replace `ClassUtils` with a temporary, local, deprecated shim that knows about `ProxyManager` proxy class names

Ocramius commented on shims/ClassUtils.php:
```diff
@@ -0,0 +1,131 @@
+<?php
```
This file is to be removed before merge

Ocramius commented:
I must say that I have NO IDEA why travis fails and local tests pass...

Ocramius committed:
Beginning work on the `Configuration` overhaul - we want to manage a `ProxyManager\Configuration` instance in the ORM config

Ocramius committed:
Setting a proxy path should update the `ProxyManager\Configuration` instance too

Ocramius committed:
Extracting temporary dir generation to a private method

Ocramius committed:
Changing auto-generation flag should also change the `ProxyManager\Configuration` referenced generator strategy

Ocramius committed:
Verifying that the `Doctrine\ORM\Configuration` builds separate and independent ghost object factory instances

Ocramius committed:
Adding `Doctrine\ORM\Configuration` API as per test specification

Ocramius committed:
`setProxyNamespace()` should also update the wrapped `ProxyManager\Configuration` instance

Ocramius committed:
`SetupTest` was using invalid proxy paths

Ocramius committed:
Rewriting `StaticProxyFactory` so its constructor just receives a `GhostObjectFactory` as constructor argument, and that's it

Ocramius committed:
Using the `LazyLoadingGhostFactory` to trigger proxy generation

Ocramius committed:
Improving performance of the `StaticProxyFactory` by caching all internal components that do not require repeated fetch/build steps

Ocramius committed:
Removing all unused proxy-related components

Slamdunk commented:
>  Replace proxy layer with ProxyManager's layer

To me sounds a lot like a big _Conflict of interest_

> I must say that I have NO IDEA why travis fails and local tests pass...

![Works on my machine](https://i.imgflip.com/9mqg4.jpg)

Ocramius commented:
Tbh, I was just sick of fixing the same bugs on both projects, plus the
proxy implementation advantages have yet to be documented 😁

On 22 Sep 2017 15:18, "Filippo Tessarotto" <notifications@github.com> wrote:

> Replace proxy layer with ProxyManager's layer
>
> To me sounds a lot like a big *Conflict of interest*
>
> I must say that I have NO IDEA why travis fails and local tests pass...
>
> [image: Works on my machine]
> <https://camo.githubusercontent.com/80febd814684733cadda8b35498500b096cc5561/68747470733a2f2f692e696d67666c69702e636f6d2f396d7167342e6a7067>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/6719#issuecomment-331444079>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakFApyepEKbDgMczjMAyzXymzT9rQks5sk7OngaJpZM4PgXck>
> .
>

Ocramius committed:
Replacing `ClassUtils` shim usage with a new `StaticClassNameConverter` with less features

Ocramius committed:
Basic tests for the `StaticClassNameConverterTest`

Ocramius committed:
Verifying all of the `StaticClassNameConverterTest` API

Ocramius committed:
Removed `ClassUtils` shim: class has been internalized

Ocramius committed:
Dead code removal

Ocramius committed:
Removing unused import

Ocramius committed:
Removing solved `TODO` in the `Parser` changes

Ocramius committed:
Removing solved TODOs and improved documentation around why we work with an un-managed proxy anyway

Ocramius committed:
Removing duplicate `TODO` that was moved to a private method documentation

Ocramius commented on lib/Doctrine/ORM/Configuration.php:
```diff
@@ -46,12 +57,16 @@ class Configuration extends \Doctrine\DBAL\Configuration
      *
      * @return string|null
+     *
+     * @deprecated please do not use this anymore
```
Should we drop this?

Ocramius replied:
Not removed: too commonly used, therefore keeping it and documented as deprecated

guilhermeblanco replied:
Drop it! Unless you can make it fully functional to keep BC.
I didn't for Metadata, so I'm assuming if people will already change for Metadata, they could do for Proxy also.

Ocramius replied:
I kept it because this is a high visibility endpoint: most integration layers with the ORM touch the configuration object. I will do a github search before considering a droo

Ocramius commented on lib/Doctrine/ORM/Configuration.php:
```diff
@@ -64,6 +79,8 @@ public function getProxyDir()
      *
      * @return int Possible values are constants of Doctrine\ORM\Proxy\Factory\StaticProxyFactory.
+     *
+     * @deprecated please do not use this anymore
```
Should we drop this?

Ocramius replied:
Not removed: too commonly used, therefore keeping it and documented as deprecated

Ocramius commented on lib/Doctrine/ORM/Configuration.php:
```diff
@@ -83,12 +100,32 @@ public function getAutoGenerateProxyClasses()
      *
      * @return string|null
+     *
+     * @deprecated please do not use this anymore
```
Should we drop this?

Ocramius replied:
Not removed: too commonly used, therefore keeping it and documented as deprecated

Ocramius committed:
Simplifying `generateProxyClasses` code

Ocramius committed:
Simplified spec for an identifier normalizer

Ocramius committed:
Completing specification for nested reference identifier normalization

Ocramius committed:
Implemented `NormalizeIdentifier` as per specification

Also made sure the `EntityManager` does not ruin identifiers when `getReference` is being called

Ocramius committed:
Using `NormalizeIdentifier` inside the `UnitOfWork` instead of relying on an additional internal method

Ocramius committed:
Minimal `NormalizeIdentifier` documentation

Ocramius committed:
Inlining variables that are only used once

Ocramius commented:
This is done so far: next up is just the documentation

Ocramius committed:
Marking utility classes as internal

Ocramius committed:
Extracting `NormalizeIdentifierTest` assets to a `Model` namespace

Ocramius committed:
Verifying that identifiers are preserved when using `EntityManager#getReference()`

Ocramius committed:
Verifying that identifier order being swapped does not cause identifier references to be lost

Ocramius committed:
Removing fixed TODO about identifier handling

Ocramius committed:
Verifying that ghost object friend objects are not initialized when not reading lazy state

Ocramius committed:
Verifying friend object "Comparable" behavior when using proxies that access each other's private state

Ocramius committed:
Using `AUTOGENERATE_EVAL` during test suite runs

Ocramius committed:
Adapting proxy configuration documentation to the new changes in the proxy handling in ORM v3

Ocramius committed:
Removing redundant/conflicting proxy documentation, correcting proxy performance/autoloading docs

Ocramius committed:
`func_get_args()` now works within proxies: YAY!

Ocramius committed:
Updating documentation about proxy serialization

Ocramius committed:
s/lazy-load/lazy-loading

Ocramius committed:
Removing FAQ entry for item that no longer holds true about `public` property lazy-loading semantics

Ocramius committed:
Removing documented quirks about proxy lazy-loading semantics combined with reflection: now always holding true

Ocramius committed:
Removing reference to previous proxy model

Ocramius committed:
Replacing generated proxy example with a stub resembling current state

Ocramius committed:
Cleaning documentation around proxy behavior

Ocramius committed:
Removing erroneous documentation about `final` methods, `c__lone` and `__wakeup` details that no longer hold true

Ocramius committed:
Marking the `StaticProxyFactory` as internal, fixing type hint documentation

Ocramius committed:
Documenting introduced BC breaks in the proxy refactoring

Ocramius committed:
Clarifying that also reflection now triggers lazy-loading

Ocramius committed:
Cleaning up performance tests before attempting a fix of the `ProxyInitializationTimeBench`

Ocramius committed:
Ignoring `phpbench.phar` and the associated public key

Ocramius committed:
Removing `NonLoadingPersister` specifics that are not to be used in benchmarks

Ocramius committed:
CMS Users must be deleted between test revolutions

Ocramius commented:
Benchmark results:

```
benchmark: UnitOfWorkComputeChangesBench, subject: benchChangeSetComputation
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 1,305.988μs             | 1,113.150μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: ProxyInitializationTimeBench, subject: benchCmsEmployeeInitialization
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 4,492.494μs             | 4,945.614μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: ProxyInstantiationTimeBench, subject: benchCmsEmployeeInstantiation
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 72,182.124μs            | 194,942.108μs                                 |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: ProxyInitializationTimeBench, subject: benchCmsUserInitialization
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 8,947.280μs             | 5,902.738μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: ProxyInstantiationTimeBench, subject: benchCmsUserInstantiation
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 119,298.386μs           | 137,659.690μs                                 |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceHydrationPerformanceBench, subject: benchHydrateAllContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 1,219.090μs             | 1,245.426μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceHydrationPerformanceBench, subject: benchHydrateFixContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 604.638μs               | 617.326μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceHydrationPerformanceBench, subject: benchHydrateFlexContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 912.946μs               | 939.842μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceHydrationPerformanceBench, subject: benchHydrateUltraContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 670.386μs               | 689.746μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: MixedQueryFetchJoinArrayHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 10,006.246μs            | 10,198.008μs                                  |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: MixedQueryFetchJoinFullObjectHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 15,528.624μs            | 15,298.930μs                                  |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: MixedQueryFetchJoinPartialObjectHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 7,344.548μs             | 7,238.552μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SimpleHydrationBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 114,677.968μs           | 112,961.742μs                                 |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SimpleInsertPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 548,459.208μs           | 518,215.356μs                                 |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SimpleQueryArrayHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 5,696.064μs             | 5,807.282μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SimpleQueryFullObjectHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 51,989.224μs            | 50,381.912μs                                  |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SimpleQueryPartialObjectHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 12,397.502μs            | 12,096.792μs                                  |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SimpleQueryScalarHydrationPerformanceBench, subject: benchHydration
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 4,043.822μs             | 4,151.792μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: ProxyInitializationTimeBench, subject: benchInitializationOfAlreadyInitializedCmsEmployees
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 2,865.794μs             | 480.390μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: ProxyInitializationTimeBench, subject: benchInitializationOfAlreadyInitializedCmsUsers
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 3,205.172μs             | 571.826μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceInsertPerformanceBench, subject: benchInsertAllContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 2,088.532μs             | 2,101.618μs                                   |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceInsertPerformanceBench, subject: benchInsertFixContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 832.212μs               | 855.418μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceInsertPerformanceBench, subject: benchInsertFlexContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 914.024μs               | 922.926μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+

benchmark: SingleTableInheritanceInsertPerformanceBench, subject: benchInsertUltraContracts
+--------+--------+------+-------------------------+-----------------------------------------------+
| groups | params | revs | vcs_branch:develop:mean | vcs_branch:cleanup/replace-proxy-factory:mean |
+--------+--------+------+-------------------------+-----------------------------------------------+
|        | []     | 5    | 946.498μs               | 942.446μs                                     |
+--------+--------+------+-------------------------+-----------------------------------------------+
```

Overall quite satisfied by these, even if initialization times times went up, you have to remember that we removed one layer of indirection for method calls, and also removed some lazy-loading for any operations not accessing internal object state.

jaapio commented on docs/en/reference/advanced-configuration.rst:
```diff
@@ -221,45 +223,76 @@ option that controls this behavior is:
+    {
+        "autoload": {
+            "psr-4": {
+                "MyProject\\": "path/to/project/sources/"
```
A , (comma) is missing at the end of this line

jaapio commented on lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -401,7 +400,7 @@ protected function hydrateRowData(array $row, array &$result)
 
                     if (! $reflFieldValue || isset($this->hints[Query::HINT_REFRESH]) ||
-                        ($reflFieldValue instanceof Proxy && ! $reflFieldValue->__isInitialized())) {
+                        ($reflFieldValue instanceof GhostObjectInterface && ! $reflFieldValue->isProxyInitialized())) {
```
It doesn't really matter, but this check is everywhere in the code. I would personally extract it to a helper function. Although an extra method call could have impact on the execution time, depending on the location of these checks.

Ocramius replied:
Sadly, this is a very common problems with the ORM: we cannot extract logic to methods without massive performance impact.

It is indeed very true that it is horrible, but php-src does not yet optimise cross-file calls.

jaapio reviewed (commented):
I'm not a doctrine insider. So I didn't understand every change. Especially the usage of final methods should be documented for me as end-user. What does it mean of my entities when I use final methods?

guilhermeblanco commented on docs/en/tutorials/getting-started.rst:
```diff
@@ -49,14 +49,14 @@ What are Entities?
 is an instance variable of the entity that is saved into and retrieved from the database
 by Doctrine's data mapping capabilities.
 
+An entity class must not be ``final``, although it can contain ``final`` methods.
```
I'm assuming some sort of Exception should be triggered in this case. If this correct, @Ocramius ?

Ocramius replied:
Yes, an exception is raised by ProxyManager when trying to generate proxies for classes that aren't really proxy-able

guilhermeblanco commented on lib/Doctrine/ORM/Proxy/Factory/ProxyFactory.php:
```diff
@@ -47,7 +47,7 @@
     const AUTOGENERATE_EVAL = 3;
 
     /**
-     * @param array<Doctrine\ORM\Mapping\ClassMetadata> $classMetadataList
```
https://i.redditmedia.com/0UfebEHWhsAa6CyTcxXnHzSnVdaw3x2TRFJjdT_-CGw.jpg?w=538&s=724ae8ed0e8391ab9f3f36c4458cb43c

Ocramius replied:
First requires to be fixed in known code analysis tools, then we can use it 😁

guilhermeblanco commented:
@Ocramius overall PR looks solid. Feel free to merge once you've looked at the comments.
The `NormalizeIdentifier` is yet for discussion, but I don't want to disrupt the main purpose of this PR for no reason.

Ocramius commented on lib/Doctrine/ORM/Configuration.php:
```diff
@@ -848,4 +887,15 @@ public function setDefaultQueryHint($name, $value)
+        return new LazyLoadingGhostFactory(clone $this->getProxyManagerConfiguration());
+    }
+
+    public function getProxyManagerConfiguration() : ProxyManagerConfiguration
```
It is mutable on purpose to allow users to customise details later on, but internally, a new copy is made every time the proxy factory is used. 

This behaviour is also tested though (to avoid the mutability bugs you mentioned)

Ocramius committed:
Adding missing trailing comma as per @jaapio's review

Ref: https://github.com/doctrine/doctrine2/pull/6719#discussion_r140642661

Ocramius commented:
@jaapio 

 > What does it mean of my entities when I use final methods?

Before this change, `final` methods in entities were NOT allowed. This limitation simply goes away :-)

Ocramius committed:
BC Break: Removing `getProxyDir`, `getProxyNamespace` and `getAutoGenerateProxyClasses` methods

Also cleans up the `Doctrine\ORM\Configuration` class, which used weird array access to save some memory (no longer true since PHP 5.4+)

Ocramius committed:
Documented removal of `getProxyDir`, `getProxyNamespace` and `getAutoGenerateProxyClasses` methods

Ocramius committed:
Making sure the `ProxyFactory#createProxy()` is always given a `ClassMetadata` instance

Ref: https://github.com/doctrine/doctrine2/pull/6719#discussion_r140646281

Ocramius committed:
Static analysis fixes as reported by scrutinizer-ci

Ocramius commented:
:ship:

carnage commented:
Done a quick scan of the diff and this seems to be a massive BC break (literally spelled out in upgrading.md); should it not be targetting 3.0.x not 2.10.x ?

greg0ire commented:
I'm AFK for a week and I don't remember what the plan was here. @beberlei, can you please advise?

Ocramius commented:
This should really land in 3.x, as the type of the proxies changes completely.

beberlei commented:
Its not really a madsive BC break as the Proxy interface is mostly internal.

The idea is, when possible and i think that might be, a user should be able to decide which proxy generator to use via config, which would allow downstreams to gradually prepare for 3.0.

That would require proxy manager to be flexible enough to also implement our Proxy interface if possible. Otherwise we can provide APIs that would enable forward compatibility
